### DISCUSSION TITLE
Фильтрация недоступных треков

### DISCUSSION THREAD

**Discussion by ilyanikolaich** [Chimildic/goofy#49] [Идея] [answered]
Можно ли как-то реализовать фильтрацию недоступных треков из плейлистов и т.д.?
Так как они обрабатываются на общих основаниях, и если, например, некий источник содержит довольно большое количество недоступных треков - то они и в отобранные плейлисты попадают, и иногда это раздражает. 
Отключить их показ в приложении(-ях) - не особо вариант, так как они  все равно добавляются в плейлисты, просто общее кол-во треков уменьшается. То есть, к примеру, отбираются 20 треков каждое утро по неким правилам, и из-за попадания в них недоступных, в реальности в плейлисте может играть всего 15-17, например.
Это конечно, не критично, так сказать, пожелание к усовершенствованию, но было бы приятно. Что-то вроде Filter.removeUnavailaible();

**Comment by Chimildic** [comment, marked as answer]
Давайте попробуем. Источников много, прошелся вроде по всем, чтобы получать данные о возможности проигрывания. Обновите код до текущего состояния. **Требуется** новое право доступа. Подробнее в [списке изменений](https://chimildic.github.io/goofy/#/changelog). О возникших ошибках пишите здесь.
```js
let tracks = Source.getPlaylistTracks('', 'id');
Filter.removeUnavailable(tracks);
```

**Reply by tupizm** [reply]
> TypeError: Cannot read property 'some' of undefined
> (функция без названия)	@ Library.gs:932
> isSomeIncludes	@ Library.gs:931
> isBelongBanGenres	@ Library.gs:927
> isBelongArtist	@ Library.gs:872
> (функция без названия)	@ Library.gs:825
> rangeTracks	@ Library.gs:824
> cleanTracks	@ 06 otherCache.gs:40
> otherCaches	@ 06 otherCache.gs:15

У меня такое посыпалось. Права обновил.

**Reply by efremych** [reply]
Здорово, что решили взяться за это дело! 
Я правильно понимаю, что теперь изменится для всех источников структура выходных данных?

Вот для такого вот трека, какой id будет использоваться? Например, при RemoveTrack.

Из консоли взял ответ, не знаю, такой ли теперь ответ с измененным кодом

 ```
 "id": "2FD1uP3rk4OIw8EbC023xB",
  "is_local": false,
  "is_playable": true,
  "linked_from": {
    "external_urls": {
      "spotify": "https://open.spotify.com/track/6kUeWuOwyHKjoACVgPOAz6"
    },
    "href": "https://api.spotify.com/v1/tracks/6kUeWuOwyHKjoACVgPOAz6",
    "id": "6kUeWuOwyHKjoACVgPOAz6",
    "type": "track",
    "uri": "spotify:track:6kUeWuOwyHKjoACVgPOAz6"
  },
  "name": "Neither Here Nor Missed",
  "popularity": 24,
  "preview_url": "https://p.scdn.co/mp3-preview/1c787f073d6159d2e28c451cc897544e8f256004?cid=774b29d4f13844c495f206cafdad9c86",
  "track_number": 5,
  "type": "track",
  "uri": "spotify:track:2FD1uP3rk4OIw8EbC023xB"
}
```

**Reply by Chimildic** [reply]
> У меня такое посыпалось

Второй раз о такой ошибки пишут. Но у меня не получается ее воспроизвести. Покажите код вашей функции `otherCaches`.

> что теперь изменится для всех источников структура выходных данных

В данный момент Spotify в треках будет давать ключ `is_playable` вместо массива стран. Других изменений нет.

**Reply by tupizm** [reply]
Чего-то не пойму как тут код красиво вставить
```js
let banTracks = Cache.read('banTracks.json');

let synthwaveMineTracks = cleanTracks(Source.mineTracks({
  keyword: ['synthwave', 'spacewave', 'popwave', 'darksynth'],
  playlistCount: 5,
  requestCount: 3,
}));
Filter.rangeTracks(synthwaveMineTracks, {
  artist: {
    genres: ['synthwave', 'spacewave', 'popwave', 'darksynth'],
  },
});

Cache.compressTracks(synthwaveMineTracks);
Cache.append('1.json', synthwaveMineTracks);
let synthwaveTracksFinal = Cache.read('1.json');
Filter.dedupTracks(synthwaveTracksFinal);
Cache.compressTracks(synthwaveTracksFinal);
Cache.write('1.json', synthwaveTracksFinal);
console.log('1', synthwaveTracksFinal.length);

function cleanTracks(tracksArray) {
  Filter.dedupTracks(tracksArray);
  Filter.removeTracks(tracksArray, banTracks);
  Filter.match(tracksArray, 'cover|live', true);
  Filter.matchLatinOnly(tracksArray, '^[a-zA-Z0-9]+', true);
  Filter.rangeTracks(tracksArray,
    {
      artist:
      {
        ban_genres:
          ['rap', 'r-n-b', 'hip-hop', 'latin', 'latino']
      }
    }
  );
  return (tracksArray);
}
```

**Reply by efremych** [reply]
> В данный момент Spotify в треках будет давать ключ `is_playable` вместо массива стран. Других изменений нет.

Получается, другой `id` трека будет приходить? Или `id `переносится из `linked_from`?

**Reply by Chimildic** [reply]
> Получается, другой `id` трека будет приходить?

Если трек недоступен, но есть аналог по `track.id` будет от аналога, а `track.linked_from.id` от оригинала. 
Если такие аналоги будут досаждать, вероятно в фильтре можно их отбрасывать. Для чего еще пригодится оригинальное `id` я пока не понимаю.

@tupizm обновите код, ошибка должна уйти

**Reply by efremych** [reply]
> Для чего еще пригодится оригинальное `id`

Фильтровать треки по `id`, например. Но даже я вроде бросил так фильтровать. Может и правда не нужен...

**Reply by efremych** [reply]
Обновил. Суть моего кейса в чём.
Если у меня в плейлисте или в любимых треках есть трек, который по задумке Spotify мне недоступен (существует аналог по релинку). То я свой трек никак не получу.
Например, трек этот у меня в любимых. Запрашиваю любимые треки, сохраняю в плейлист, в итоге трек в плейлисте без сердечка.

Надо подумать, может получится перезаписывать `id` трека. Или смириться.

**Reply by Chimildic** [reply]
Если трек в любимых и стал недоступен, он никуда не девается. Просто при создании плейлиста будет аналог или совсем не будет такого трека (при использовании фильтра). Как описывает первый пост, не занимает место другого трека.
Нужен ли аналог пока вопрос открытый. Я не знаю насколько они точные. Может там одинаковые альбомы, но под разные страны. А может аналогом вдруг ремикс считается.

**Reply by ilyanikolaich** [reply]
Интересно, что если получить массив треков чтением из кеша, например 
`let SavedTracks = Cache.read('Liked.json');` то `Filter.removeUnavailable(SavedTracks);` не отрабатывает - остается то же самый массив, который и был изначально, но если `SavedTracks ` сохранить в свой плейлист, и потом получить треки из этого сохраненного плейлиста, они уже обрабатываются `Filter.removeUnavailable(SavedTracks);` вполне должным образом.

**Reply by tupizm** [reply]
А я думал это я криворукий и не получается сделать, поэтому забил. Оказывается система

**Reply by Chimildic** [reply]
@ilyanikolaich чтобы фильтровать, нужен ключ `is_playable`. Ваш кэш с предыдущих версий его не содержит. Поэтому проходят все треки. При новом запросе треков от Spotify, работает потому что к запросу добавляется необходимость добавить такой ключ. Теоретически, можно пересохранить кэш новым набором треков. Но значение `is_playable` будет тем больше неактуальным, чем дольше трек в кэше.

**Reply by tupizm** [reply]
А как его можно пересохранить с этим ключом?

**Reply by Chimildic** [reply]
@tupizm добавьте треки кэша в плейлист, запросите этот плейлист и пересохраните кэш. Но еще раз, значение будет неактуальным с течением времени. По опыту Яндекса, права приходят-уходят каждый день.
```js
let cacheTracks = Cache.read('file.json');
Playlist.saveAsNew({ name: 'кэш', tracks: cacheTracks, });
let tracks = Source.getPlaylistTracks('кэш');
Cache.compressTracks(tracks);
Cache.write('file.json', tracks);
```

**Reply by tupizm** [reply]
Понятно, то есть проще всего отсеивать на этапе создания плейлиста, как я понял.

**Reply by Chimildic** [reply]
Зависит от сценария использования кэша. Если он для остальной фильтрации, не нужно ничего делать. Если его треки попадают в плейлист - посмотреть как часто и как много он приводит недоступных треков.

**Reply by tupizm** [reply]
У меня кеш чисто, как хранилище используется. Так что мне подойдёт вариант непосредственно перед плейлистом. Кстати, в обнове последней не было чего-то связанно с жанрами? У меня лежит кеш, там собираю классику рока с помощью майнера и отбором по жанру. Давно он у меня плавал в размере 7тыс треков. Вчера запустил скрипт на 1.4.4 и сразу 10тыс стало.

**Reply by efremych** [reply]
Пока вот так себе оставлю, если ошибок не будет и если это единственное место, которое надо поправить.

  ``` js
  function extractTracks(items) {
      if (!items || items.length == 0) return [];
        let key = items[0].played_at ? 'played_at' : 'added_at';        
        return items.reduce((tracks, item) => {        
            if ((!item.hasOwnProperty('is_local') || !item.is_local) && item.track && item.track.artists && item.track.artists.length > 0) {
                let date = item[key] ? item[key] : DEFAULT_DATE.toISOString();
                item.track[key] = date;
                if (item.hasOwnProperty('origin')) {
                    item.track.origin = item.origin;
                }
                if (item.track.hasOwnProperty('linked_from')) { 
                    item.track.id = item.track.linked_from.id;
                    item.track.external_urls.spotify = item.track.linked_from.external_urls[0];
                    item.track.href = item.track.linked_from.href;
                    item.track.uri = item.track.linked_from.uri;
                }
                tracks.push(item.track);
                item.track.source = item.source;
            }
            return tracks;
        },
        []);
    }
```

Чтобы результат был таким при сохранении в Плейлисте№2 треков из Плейлиста№1
![Безымянный](https://user-images.githubusercontent.com/76212904/107844881-62102e80-6de8-11eb-8432-8ee4e8134414.jpg)

А не таким
![Безымянный2](https://user-images.githubusercontent.com/76212904/107844885-6b010000-6de8-11eb-8ed9-ab55bbca5d80.jpg)

**Reply by Chimildic** [reply]
@efremych если трек доступен, откуда у него появляется `linked_from`?

@tupizm исправлялась ошибка, с которой вы столкнулись ранее. Результат отбора не соответствует заданным значениям?

**Reply by tupizm** [reply]
> Результат отбора не соответствует заданным значениям?

А бог его знает. Я даже не знаю, как это проверить в таком количестве треков 😀 у меня так все кеши скакнули. Уже думал заново начать майнить, чтобы очистить это дело все.

**Reply by efremych** [reply]
> если трек доступен, откуда у него появляется `linked_from`

Так потому и доступен (можно проиграть), потому что переадресуется на друой трек. Следовательно, "linked_from"

Трек №1 - залайкан или просто хранится у меня в библиотеке (id = `6kUeWuOwyHKjoACVgPOAz6`).
Данный трек нельзя найти через Поиск в Spotify. Так как альбом недоступен в моей стране.
Каким-то образом удалось заполучить данный трек/залайкать. Теперь вот лежит.
_кстати, иногда удобно иметь такой трек, чтобы открыть недоступный альбом (ну, зачем-нибудь)_
```
"id": "2FD1uP3rk4OIw8EbC023xB",
  "is_local": false,
  "is_playable": true,
  "linked_from": {
    "external_urls": {
      "spotify": "https://open.spotify.com/track/6kUeWuOwyHKjoACVgPOAz6"
    },
    "href": "https://api.spotify.com/v1/tracks/6kUeWuOwyHKjoACVgPOAz6",
    "id": "6kUeWuOwyHKjoACVgPOAz6",
    "type": "track",
    "uri": "spotify:track:6kUeWuOwyHKjoACVgPOAz6"
  },
```

Трек№2 - трек, который Spotify мне будет "подсовывать" вместо Трека №1 (например, в Истории прослушивания)
Альбом, с которого данный трек, доступен в моей стране и трек можно найти в Поиске в Spotify.
Никаких `linked_from` у него нет
```
"id": "2FD1uP3rk4OIw8EbC023xB",
  "is_local": false,
  "is_playable": true,
```

Бонус, Трек №3 Просто недоступный трек. С ним всё понятно
```
 "id": "395b9mfjEHhWfb9YAWMmsl",
  "is_local": false,
  "is_playable": false,
  "name": "Tomorrow",
```

**Reply by Chimildic** [reply]
@efremych по скриншоту, у лайка не видно серого цвета. Я только такие недоступные видел. Какой в нем смысл? Если нельзя проиграть. 

@tupizm возможно рост связан с исправлением ошибки. Она касалась использования `rangeTracks` подряд. 
В вашем скрипте достаточного и одного
```js
  Filter.rangeTracks(synthwaveMineTracks, {
    artist: {
      genres: ['synthwave', 'spacewave', 'popwave', 'darksynth'],
      ban_genres: ['rap', 'r-n-b', 'hip-hop', 'latin', 'latino']
    },
  });
```

**Reply by efremych** [reply]
> по скриншоту, у лайка не видно серого цвета. Я только такие недоступные видел. Какой в нем смысл? Если нельзя проиграть.

Это "Трек №1". Его можно проиграть, абсолютно нормальная сущность. Но.
В Историю прослушивания попадет Трек №2, в рекомендации Spotify также попадет Трек№2 (хотя Spotify знает, что у меня в любимых трек с другим айди)

Все эти мои, возможно лишние, телодвижения, потому что я не хочу перелайкивать трек, меня и этот устраивает. И хочу, чтобы когда я его добавлю в плейлист с помощью скрипта, этот трек был с сердечком.

Возможно, еще подумаю, чтобы в Общую Историю прослушивания также попадал именно Трек№1, а не Трек№2. Но это не точно.

**Reply by Chimildic** [reply]
@efremych  Доступен трек, но недоступен альбом и поэтому линк на другой трек? Спотифай удивляет все больше. 
Не проще перелайкать трек и забыть о костылях?

**Reply by efremych** [reply]
> Не проще перелайкать трек и забыть о костылях?

Проще, конечно.
Но если есть такая возможность, почему б не воспользоваться.

Есть у меня в любимых, например, трек с недоступного в моей стране альбома, в котором, условно, 30 треков.
Spotify мне предлагает такой же трек, но с альбома на 15 треков. Оставшиеся 15 треков больше никак не найти (они реально недоступны для проигрывания).
Так что трек, который у меня сохранен и доступен для проигрывания является "ярлыком" к этому альбому на 30 треков. Мол, "вот у исполнителя есть еще вот такие треки, хочешь найди на стороннем ресурсе, ознакомься с этими треками" .

Не удивлюсь, если Трек№1 и Трек№2 могут отличаться качеством. Но это маловероятно.

**Comment by Chimildic** [comment]
Что делать с релинком трека когда недоступен оригинал? То есть подмена на доступный аналог.

Исходя из описания в предыдущем посте, возможна ситуация когда: трек доступен, альбом недоступен - Spotify делает релинк. В результате, например, трек с лайком заменится на аналог без лайка. С другой стороны, при различном поиске, рекомендациях - релинк скорее полезен. 

В текущем варианте, к `removeUnavailable` можно добавить аргумент, по которому принимается решение о использовании релинка или оригинала. Но тогда это заставляет использовать `removeUnavailable` везде, где хочется уйти от релинка. 

Альтернатива, откатиться назад. Не получать ключ `is_playable`, не получать релинк ни в каком виде. Принимать решение о доступности трека через ключ `available_market`, проверяя на наличие значения `RU`. 

Оба варианта никак не решают ситуацию с кэшем. Поскольку `compressTracks` удаляет `available_market`, а то что добавлялось без сжатия может быть также неактуальным как и в случае с `is_playable`. И здесь можно пойти дальше, потратив запросы на перепоручение полных объектов для этих треков, чтобы иметь `is_playable` или `available_market`. Но стоит ли это делать? Потому что `removeUnavailable` логично использовать до остальных фильтров, если следовать тезису из основного поста про занимаемое место. Без остальных фильтров, треков вначале может быть много. Хотя их и можно переполучить пачками по 50 на 1 запрос. В результате кэш на 10 тысяч треков отправит 200 запросов.

**Reply by efremych** [reply]
В текущей реализации вопрос только в наличии лайка на выходе (кому-то может и пофиг) + больше шансов дубликат имеющегося трека добавить себе в библиотеку. Ну, и по `id` трека не пытаться отсеивать треки (`Filter.RemoveTracks`).
Как вариант, добавить функцию, которая пройдется по библиотеке при желании юзера и заменит релинкнутые треки на доступные (снимет/добавит лайки, удалит/добавит в соответствующий плейлист).
Так  сказать, чтобы юзер шёл в ногу со Спотифаем)



> к `removeUnavailable` можно добавить аргумент, по которому принимается решение о использовании релинка или оригинала.

А тут зачем, `removeUnavailable` же отсеивает треки `is_playable = false`? Это же и нужно от функции. Если `false`, релинка же не будет, ну, или я таких треков не видел.



> доступности трека через ключ `available_market`, проверяя на наличие значения `RU`

Я так на первых порах хотел, но так как по большей части, меня интересовал анализ своей библиотеки на предмет недоступных треков, то отсеял этот вариант - так как у релинкнутых треков `available_market` не содержит `RU`



> Оба варианта никак не решают ситуацию с кэшем

Только понять и принять, на мой взгляд. Максимум, как то массово обновлять кэш периодически... Но тогда теряется смысл кэша.
Иначе, можно было бы реализовать и мою мысль про оптимизацию запросов к Spotify, путем сохранения поисковых запросов и их результатов в кеше (для LastFM, Radio и т.п.)



> Что делать с релинком трека когда недоступен оригинал

А костыль, который я себе придумал, не подходит? По идее он возвращает нас к исходной точке.
Хорошо потестить пока нет возможности, но вроде логично.

**Comment by Chimildic** [comment]
Полнейшая неразбериха. На своих лайках получил такое поведение:

Трек можно послушать
- в `available_markets` есть `RU`
- в `available_markets` нет `RU`, но есть релинк - оригинал доступен (!)
- `available_markets` пустой, есть релинк - оригинал доступен (!)

Трек нельзя послушать
- в `available_markets` нет `RU`, нет релинка
- `available_markets` пустой, нет релинка

Но ведь возможна ситуация, когда нет `RU` или пустой массив, есть релинк и оригинал недоступен. Как это тогда вообще определять? 

[Обновил код](https://github.com/Chimildic/goofy/blob/main/library.js), чтобы проверять такие состояния. Убрал запросы на `is_playable`, релинков нет. Кто-нибудь найдет случай описанный выше? То есть после фильтра в нем окажется недоступный трек.

**Reply by efremych** [reply]
> Но ведь возможна ситуация, когда нет `RU` или пустой массив, есть релинк и оригинал недоступен

А почему такая ситуация должна быть возможна? Потому что Spotify?

**Reply by Chimildic** [reply]
@efremych наличие релинка должно говорить о недоступности трека. Но он доступен. Тогда зачем релинк? Выходит все таки есть случай недоступного оригинала с аналогом. Тогда как отличить одного от другого? Если ни `available_markets`, ни `is_playable`, ни `linked_from` в этом не помогают.

**Reply by efremych** [reply]
Насколько я понимаю, возможны только 3 ситуации:
-  `is_playable` = `false` однозначно говорит, что трек недоступен - никак не прослушать. Релинка в этой ситуации нет.
- `is_playable` = `true`  + нет релинка. Трек с альбома, доступного в `RU`. И трек и альбом можно найти в поиске.
- `is_playable` = `true`  + релинк. Трек с альбома, НЕдоуступного в `RU`. И трек и альбом НЕможно найти в поиске. В релинке `id` нашего трека, а сам ответ содержит данные трека-аналога, доступного в `RU`

> наличие релинка должно говорить о недоступности трека. Но он доступен. Тогда зачем релинк?

Формально трек недоступен, его же не найти в базе Spotify. Наверно, сделано для удобства пользователей, которые сохранили у себя трек, а альбом позже стал недоступным, потому что релизнулась другая его версия в регионе. Куча негатива бы была, а так, формально, Spotify выпилил трек из данного рынка. Иначе пользователю бы пришлось чуть ли не каждый день треки заново искать.

Странно тогда, что Spotify принудительно не обновляет трек у пользователя.
У меня так было с треком. у которого есть USA версия и International версия. Еще в самом начале появления Spotify в России. Я сохранил у себя USA версию (другой не было). А через некоторое время мне принудительно заменили трек на International версию (они отличаются звучанием, можно это заметить, плюс файл трека перезакачивался) .

Обновил код, вроде б работает `removeUnavailable`

**Reply by Chimildic** [reply]
Еще раз перечитал [документ про релинк](https://developer.spotify.com/documentation/general/guides/track-relinking-guide/). 

> when a user tries to play a track that isn’t available in their own market, the Spotify mobile, desktop, and web players try to play another instance of the track that is available in the user’s market.

То есть на ходу подменяют оригинал релинком для прослушивания. Но при этом игнорируют лайк когда встречаешь сам релинк.

**Reply by efremych** [reply]
А в [Get Current User's Recently Played Tracks](https://developer.spotify.com/documentation/web-api/reference/#endpoint-get-recently-played), похоже лайков и не получить, если только не через 10 кругов ада :(

**Reply by Chimildic** [reply]
@efremych получается, что мы слушаем релинк. Поэтому они в историю и возвращаются.

**Reply by Chimildic** [reply]
@efremych посмотрел, даже с указанием `market`, ответ не содержит `linked_from`. Поэтому оригинал нам не подставить.

**Comment by amd64fox** [comment]
Получил такую ошибку, которая остановила скрипт.
```
Ошибка	
TypeError: Cannot read property 'id' of undefined
(функция без названия)	@ Library.gs:971
removeUnavailable	@ Library.gs:969
Povtori	@ Main.gs:2159
```

Main 2159  `Filter.removeUnavailable(moi_mix);`

Библиотеку обновил. права дал, проверял это фильтр на тестовом примере небольшом, там отработал без ошибок.

**Reply by efremych** [reply]
Еще `getArtistsTopTracks` похоже нужно вернуть `country` - ругается.

У `getArtistsAlbums` тоже раньше было `country` - но вроде работает и так.

**Reply by Chimildic** [reply]
@amd64fox @efremych обновил
